tibco_ems/datadog_checks/tibco_ems/tibco_ems.py

```diff
@@ -0,0 +1,247 @@
+DEFAULT_PORT = 7222
+TO_BYTES = {'B': 1, 'Kb': 1e3, 'Mb': 1e6, 'Gb': 1e9, 'Tb': 1e12}
+CONNECTION_STRING = 'tcp://{}:{}'
+
+
+class TibcoEMSCheck(AgentCheck):
+
+    # This will be the prefix of every metric and service check the integration sends
+    __NAMESPACE__ = 'tibco_ems'
+
+    def __init__(self, name, init_config, instances):
+        super(TibcoEMSCheck, self).__init__(name, init_config, instances)
+        # Allow to specify a complete command for tibemsadmin such as `docker exec <container> tibemsadmin`
+        default_tibemsadmin_cmd = init_config.get('tibemsadmin', 'tibemsadmin')
+        tibemsadmin_cmd = self.instance.get('tibemsadmin', default_tibemsadmin_cmd).split()
+        host = self.instance.get('host', DEFAULT_HOST)
+        port = self.instance.get('port', DEFAULT_PORT)
+        username = self.instance.get('username')
+        password = self.instance.get('password')
+        script_path = self.instance.get('script_path')
+        server_string = CONNECTION_STRING.format(host, port)
+        self.tags = self.instance.get('tags', [])
+        self.parsed_data = {}
+
+        self.cmd = tibemsadmin_cmd + [
+            '-server',
+            server_string,
+            '-user',
+            username,
+            '-password',
+            password,
+            '-script',
+            script_path,
+        ]
+
+    def check(self, _):
+
+        output, err, code = self.run_tibco_command()
+        if err or 'Error:' in output or code != 0:
+            self.log.error('Error running command: %s', err)
+            return
+
+        # Sanitize the output
+        cleaned_data = output.replace('\r', '').strip()
+
+        # Split the output into command sections
+        sections = self._section_output(cleaned_data)
+
+        # Parse the output
+        for command, section in sections.items():
+            pattern = SHOW_METRIC_DATA[command]['regex']
+            if command == 'show server':
+                self.parsed_data[command] = self._parse_show_server(section, pattern)
+            else:
+                try:
+                    self.parsed_data[command] = self._parse_factory(section, pattern)
+                except Exception as e:
+                    self.log.error('Error parsing command %s: %s', command, e)
+                    continue
+
+        for command, metric_info in self.parsed_data.items():
+            metric_keys = SHOW_METRIC_DATA[command]['metric_keys']
+            tag_keys = SHOW_METRIC_DATA[command]['tags']
+            metric_prefix = SHOW_METRIC_DATA[command]['metric_prefix']
+
+            if command == 'show server':
+                self._submit_metrics_factory(metric_prefix, metric_info, metric_keys, tag_keys)
+            else:
+                for metric_entry in metric_info:
+                    self._submit_metrics_factory(metric_prefix, metric_entry, metric_keys, tag_keys)
+
+    def run_tibco_command(self):
+        return get_subprocess_output(self.cmd, self.log)
```

Comment:
nit. I think it would be good to debug log or trace log the self.cmd for troubleshooting purposes.

Reply:
Hmm `self.cmd` does get automatically logged in debug mode as part of the `get_subprocess_output` module but it doesn't obfuscate passwords. That's an adjustment that can be made in the base class 👍